// react-router-loading/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LoadingContext", {
  enumerable: true,
  get: function get() {
    return _LoadingContext.LoadingContext;
  }
});
Object.defineProperty(exports, "LoadingGetterContext", {
  enumerable: true,
  get: function get() {
    return _LoadingContext.LoadingGetterContext;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _LoadingSwitch["default"];
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function get() {
    return _LoadingRoute["default"];
  }
});
exports.topbar = void 0;

var _LoadingContext = require("./LoadingContext");

var _LoadingSwitch = _interopRequireDefault(require("./LoadingSwitch"));

var _LoadingRoute = _interopRequireDefault(require("./LoadingRoute"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var topbar = typeof window === 'undefined' ? {
  show: function show() {},
  hide: function hide() {},
  config: function config() {}
} : require('topbar');
exports.topbar = topbar;
